src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,759 @@
+            {
+                return _backup.Remove(key);
+            }
+
+            if (Count == 0)
+            {
+                return false;
+            }
+
+            var index = FindIndex(key);
+            if (index >= 0)
+            {
+                _count--;
+                var array = _arrayStorage;
+                Array.Copy(array, index + 1, array, index, _count - index);
+                array[_count] = default;
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to remove and return the value that has the specified key from the <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="key">The key of the element to remove and return.</param>
+        /// <param name="value">When this method returns, contains the object removed from the <see cref="SmallCapacityDictionary{TKey, TValue}"/>, or <c>null</c> if key does not exist.</param>
+        /// <returns>
+        /// <c>true</c> if the object was removed successfully; otherwise, <c>false</c>.
+        /// </returns>
+        public bool Remove(TKey key, out TValue? value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+
+            if (_backup != null)
+            {
+                return _backup.Remove(key, out value);
+            }
+
+            if (_count == 0)
+            {
+                value = default;
+                return false;
+            }
+
+            var index = FindIndex(key);
+            if (index >= 0)
+            {
+                _count--;
+                var array = _arrayStorage;
+                value = array[index].Value;
+                Array.Copy(array, index + 1, array, index, _count - index);
+                array[_count] = default;
+
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to the add the provided <paramref name="key"/> and <paramref name="value"/> to the dictionary.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns>Returns <c>true</c> if the value was added. Returns <c>false</c> if the key was already present.</returns>
+        public bool TryAdd(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_backup != null)
+            {
+                return _backup.TryAdd(key, value);
+            }
+
+            if (ContainsKeyCore(key))
+            {
+                return false;
+            }
+
+            EnsureCapacity(Count + 1);
+            _arrayStorage[Count] = new KeyValuePair<TKey, TValue>(key, value);
+            _count++;
+            return true;
```

Comment:
Is this code assuming that increase in capacity hasn't switched to using backup dictionary?

Reply:
Yes. If the backup dictionary is not null, it will always use it.